BUG: fix compatibility with numpy 2.0 copy semantics

neutrinoceros opened the pull request:
### Description
This adresses an new incompatibility with numpy dev:
`np.array(..., copy=False)` now raises a `ValueError` if a copy cannot be avoided. The suggested replacement, [implemented in scikit-learn by the author of the breaking change](https://github.com/scikit-learn/scikit-learn/pull/28323), is to use `np.asarray(...)` instead, which will only create a copy if necessary, as desired.

xref https://github.com/numpy/numpy/pull/25168


Note that this doesn't fix devdeps by itself: there's also [an upstream issue in pyerfa](https://github.com/liberfa/pyerfa/issues/133),~ and another incompat with `Unit` that I'll need to bisect, and which I think can be addressed separately.~
update: turns out this other incompatibility was an artifact of my first attempt at this patch, not an actual issue.

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

neutrinoceros commented:
Whoops, the whole CI just broke with
```python-traceback
TypeError: unsupported operand type(s) for *: 'Unit' and 'float'
```
Incidentally, this is the "other error" that I saw as I tested against numpy dev locally, but I didn't realise that it was actually caused by this patch. So the good news is that we probably have only one problem to solve here ! The bad news is that it's worth than I thought.

rgommers commented:
> The bad news is that it's worth than I thought.

This was an unintended regression in `numpy`; it should be emitting a warning if `copy` is missing from the signature rather than raising a `TypeError`. Should be fixed within the next 24 hours (hopefully sooner). Upstream issue is https://github.com/numpy/numpy/issues/25916.

neutrinoceros commented:
Thanks for jumping in @rgommers !
Actually I wasn't talking about this specific aspect, I just got confused when I blindly replaced `np.array(..., copy=False)` with `np.asarray(...)` without realising that some arguments (`ndmin`, `subok`) don't make sense there.

As for the TypeError VS warning, it doesn't make a difference on our side since warnings are treated as errors in CI, and in both case it's very useful information. Still, a warning would indeed be nicer for end users. Any way, thank you for all your work on numpy 2 !

mhvk commented on astropy/table/column.py:
```diff
@@ -522,6 +522,10 @@ def __new__(
         copy=False,
```
So, this could become `COPY_IF_NEEDED` if we put it in `numpycompat`

mhvk reviewed (commented):
Thanks! Looking through the changes,I think two things would simplify this:
1. I think we should define `COPY_IF_NEEDED = False if NUMPY_LT_2_0 else None` in `numpycompat` and just import that.
2. The `__array__` implementations can all have `copy=COPY_IF_NEEDED` as an argument, independent of numpy version.

rgommers commented:
> I think two things would simplify this:

I'll note that that is pretty much exactly what we did in SciPy: https://github.com/scipy/scipy/pull/20172. Looks pretty clean that way.

neutrinoceros commented on astropy/modeling/tests/test_quantities_model.py:
```diff
@@ -118,9 +118,10 @@ def _allmodels():
+            except (TypeError, IndexError, ValueError):
                 pass
-            allmodels.append(m)
+            else:
```
Most likely yes, for now I'm just trying to survive CI, and I needed that at some point along the way but I intend to remove it once the patch is stable.

neutrinoceros replied:
dropped !

neutrinoceros commented:
Thanks @mhvk ! Right now I'm trying to eliminate any regressions seen in CI but I'll get to your suggestions next !

neutrinoceros commented:
One regression to go before refactoring. Not sure I'll have time to finish this today but I should see the end of it tomorrow.

neutrinoceros commented on astropy/convolution/core.py:
```diff
@@ -177,11 +177,20 @@ def __rmul__(self, value):
-        Array representation of the kernel.
-        """
-        return self._array
+    if NUMPY_LT_2_0:
```
in hindsight, yes. I'll simplify this.

neutrinoceros commented on astropy/convolution/core.py:
```diff
@@ -177,11 +177,20 @@ def __rmul__(self, value):
+            """
+            Array representation of the kernel.
+            """
+            return self._array
```
I've added this to my personal TODO list and will make sure to address it in a follow up PR. Thanks !

neutrinoceros replied:
also reported in #16168

neutrinoceros commented on astropy/modeling/polynomial.py:
```diff
@@ -533,7 +533,7 @@ def fit_deriv(self, x, *params):
             The Vandermonde matrix
         """
-        x = np.array(x, dtype=float, copy=False, ndmin=1)
+        x = np.atleast_1d(np.asarray(x, dtype=float))
```
you got it !

neutrinoceros replied:
`COPY_IF_NEEDED` sounds like a good idea idea !

However it's no use here: `np.asarray` doesn't have a `copy` argument in numpy 1.x, but the default behaviour matches `copy=COPY_IF_NEEDED` in all versions.

mhvk replied:
I'd just prefer to keep the `np.array` call instead of effectively doing `np.array` twice (via `np.asarray` and `np.atleast_1d`).

neutrinoceros replied:
Oh, ok, sorry I misunderstood your suggestion !

neutrinoceros commented:
I'm not done refactoring, but I'm going to leave this in a functioning state for today.

neutrinoceros commented:
update:
- I rebased onto #16147 to keep CI relevant
- I've started replacing my "band aid" patched with widespread internal use of `COPY_IFF_NEEDED`, but it's taking a while to go through all of them. The last band aid is still there and only partially made obsolete.

I'm hoping I can finish this later today.

neutrinoceros commented:
Refactor is expected to be over (unless a new regression appears in CI). Next, I'll clean the branch history.
Can anyone set a milestone (presumably 6.0.1) and update labels manually ? I think we're missing:
- time
- coordinates
- utils
- no-changelog

neutrinoceros commented:
Seems like I made a mistake at some point and this is still not working with numpy 2. I'll bisect my branch while I can (will squash later).

neutrinoceros commented:
I simplified the branch history but left some of the granularity since it may be helpful for reviewers to understand my process. Ultimately this should be squashed (on merge, or before that if requested). The only remaining failure is unrelated and fixed in #16150

mhvk reviewed (commented):
This looks good modulo my wish for just using IF instead of IFF, and a few places where I think I'd pick a different choice.

But a larger point is that I was rather shocked at how far this propagates! And it is clear we need to make a decision: do we just follow numpy and insist that also for `Representation`, `Quantity`, and `Masked` people have to change their code to use `None` instead of `False`, or do we stick with how it was?

I think since we are generally passing on to numpy, what you've done is the right choice in principle: change our default for `copy` so that it matches what we want to do. However, this is an API change, since now if other people use `Quantity(..., copy=False)`, it no longer has the same meaning on numpy 2.0! So, we somehow need to deprecate that, which is actually hard since what exactly is our deprecation message going to be? Right now, for numpy >= 2.0, our default will become `None`, so we can recognize it if someone passes in `False` and just turn it into `None`. But what if they actually want the new meaning of `False`? They would not be able to indicate that easily. Perhaps that is fine, though: we can have `AstropyDeprecation` warning for `False` and `numpy>=2.0`, which tells about `COPY_IF_NEEDED` and only when our minimum version is 2.0 make the API change.

Now another question is whether we should do the deprecation in *this PR*. My sense is that here we just want to unblock CI and we should not be changing API at all. So, I *think* that implies we should not change defaults in any public methods/constructors (but rather have some form of `if not NUMPY_LT_2_0 and copy is False: copy = None`). Which then in a follow-up PR we can change (and not backport; maybe even for 7.0 only).

neutrinoceros commented on astropy/coordinates/tests/test_representation.py:
```diff
@@ -2208,19 +2209,20 @@ def test_roundtrip(self, item):
+def test_differential_norm_noncartesian(cls, ndim):
     # The norm of a non-Cartesian differential without specifying `base` should error
-    rep = cls(0, 0, 0)
+    args = tuple(0 for _ in range(ndim))
```
This worked "by accident" because the third `0` would be incorrectly used as `copy` but would simply be treated as falsy down the line :)

neutrinoceros commented on astropy/coordinates/angles/core.py:
```diff
@@ -15,7 +15,7 @@
 from astropy.units import SpecificTypeQuantity
 from astropy.utils import isiterable
-from astropy.utils.compat import NUMPY_LT_2_0
+from astropy.utils.compat import COPY_IFF_NEEDED, NUMPY_LT_2_0
```
At first I went with `COPY_IF_NEEDED` as you suggested and then got confused after a couple hours spent applying it everywhere needed 😅 
"iff" stands for "if and only if", but I agree that jargon should be avoided whenever possible, so how about `COPY_IF_AND_ONLY_IF_NEEDED`, or `COPY_ONLY_IF_NEEDED`, or `COPY_IF_YOU_MUST` (or `ONLY_SITHS_DEAL_IN_ABSOLUTES` ?) ?

mhvk replied:
I guess something like `avoid_copy` might also work, but we should not overthink this. And it looks like scipy went with `copy_if_needed`, so there is precedent...

neutrinoceros replied:
ok, let's just follow Scipy then !

neutrinoceros replied:
(though to make my life easier regarding history rewriting, I'll postpone that to after we agree on a strategy !)

neutrinoceros commented:
>  However, this is an API change, since now if other people use Quantity(..., copy=False), it no longer has the same meaning on numpy 2.0!

I'd be 50% on this point: astropy has never been used with numpy 2.0 (yet). From my stand point we're breaking the API no matter what we do because our implicit contract with our users is that
1)`copy=False` means "avoid copying if you can"
2) our API follows that of numpy

and we cannot maintain both promises at the same time... but then again maybe we can (see after) ?

> So, we somehow need to deprecate that, which is actually hard since what exactly is our deprecation message going to be? Right now, for numpy >= 2.0, our default will become None, so we can recognize it if someone passes in False and just turn it into None. But what if they actually want the new meaning of False? They would not be able to indicate that easily. Perhaps that is fine, though: we can have AstropyDeprecation warning for False and numpy>=2.0, which tells about COPY_IF_NEEDED and only when our minimum version is 2.0 make the API change.

The only way I can think of that we can accomodate everyone would be to introduce some configuration option or env variable (both ?) that would expose the choice between "legacy" behaviour and future numpy behaviour to the user (we would still need to decide wether this option should be an opt-in for future behaviour or opt-out for legacy behaviour. I think I would lean toward the latter).

> Now another question is whether we should do the deprecation in this PR. My sense is that here we just want to unblock CI and we should not be changing API at all. So, I think that implies we should not change defaults in any public methods/constructors (but rather have some form of if not NUMPY_LT_2_0 and copy is False: copy = None). Which then in a follow-up PR we can change (and not backport; maybe even for 7.0 only).

I think my first commit, isolated, would already be close to your suggested backport.
Maybe we could move forward in steps:
- keep my "band aid" patches in this PR, revert all signature changes -> backport
- in 6.1.0, add deprecation warnings were "band aid" are (with all internal changes being done here, we know this won't create self-triggered warnings)
- in 7.0 (or 8.0 ?) end these deprecations (and deprecate the option I proposed, if this part reaches consensus)

mhvk commented on astropy/table/column.py:
```diff
@@ -519,9 +519,13 @@ def __new__(
         format=None,
         meta=None,
-        copy=False,
+        copy=COPY_IFF_NEEDED,
```
For the "band aid" solution, I think this would need to remain `False`, with the below band-aid patch remaining in place.

mhvk commented on astropy/units/quantity.py:
```diff
@@ -541,6 +541,10 @@ def __new__(
                 elif unit is not value_unit:
                     copy = False  # copy will be made in conversion at end
 
+        if not NUMPY_LT_2_0 and (copy is False or copy == 0):
```
For the band aid, I think just `copy is False` is good enough.

mhvk reviewed (commented):
@neutrinoceros - sounds like we have a plan at least for step 1! I just checked your first commit and agree that would be close to what's needed to get CI moving again (and set us up nicely for the next steps). Only need to change the default for `Column` (all the changes for `__array__` are OK since those are new in 2.0 anyway).

neutrinoceros commented:
I've reset the branch to a more minimal state that should be backportable. Discarded changes will be re-submitted in a following PR. I also plan to open a separate issue to discuss out long term strategy, however, as I'm away tomorrow, I won't be able to do that before Wednesday, so anyone is welcome to beat me to it (but I'm assuming it can wait for a couple days).

mhvk commented:
Remaining failure in devdeps will be fixed by #16150

neutrinoceros commented:
(note: I'm not sure if all labels are still relevant now)

mhvk commented on astropy/units/quantity.py:
```diff
@@ -445,7 +445,12 @@ def __new__(
+        if not NUMPY_LT_2_0 and copy is False:
+            # strict backward compatibility
+            # see https://github.com/astropy/astropy/pull/16142
+            copy = None
```
I think beyond this point any time one has `copy=False`, one should set `copy=COPY_IF_NEEDED` instead, so that we avoid doing this another time. Also, as is, l.458 would lead to a problem at l.466...

mhvk commented on astropy/units/quantity.py:
```diff
@@ -541,6 +546,11 @@ def __new__(
                 elif unit is not value_unit:
                     copy = False  # copy will be made in conversion at end
 
+        if not NUMPY_LT_2_0 and copy is False:
```
So, I'd hope to get rid of doing this twice....

neutrinoceros commented:
@mhvk I will only have very limited access to a keyboard (possibly none) until Wednesday so feel free to change this bit.

mhvk reviewed (approved):
Approving. Since this now introduces no API changes at all, i.e., just accounts for the change in `np.array` in numpy 2.0, I'll go ahead and approve and set ready to merge. @neutrinoceros will open a new issue/PR with plans for how to do this going forward.

pllim commented:
Are you sure you want to auto-merge? numpy-dev job is not required to pass before it kicks in.

pllim commented:
Hmm... Is this related?

```
2024-03-04T22:00:37.4696084Z _____________________________ test_array_api_init ______________________________[0m
2024-03-04T22:00:37.4696733Z 
2024-03-04T22:00:37.4696944Z     @pytest.mark.filterwarnings(
2024-03-04T22:00:37.4697774Z         "ignore: The numpy.array_api submodule is still experimental. See NEP 47."
2024-03-04T22:00:37.4698595Z     )
2024-03-04T22:00:37.4698990Z     def test_array_api_init():
2024-03-04T22:00:37.4699537Z >       import numpy.array_api as np_api
2024-03-04T22:00:37.4700464Z E       ModuleNotFoundError: No module named 'numpy.array_api'[0m
2024-03-04T22:00:37.4701099Z 
2024-03-04T22:00:37.4702554Z astropy/units/tests/test_quantity_array_methods.py[0m:649: ModuleNotFoundError
```

MridulS commented:
@pllim that should be fixed with https://github.com/astropy/astropy/pull/16150

mhvk commented:
I knew it was going to fail given #16150, but indeed, given that that is the simpler PR, might as well check here that we are now all green again. So, will rebase and then let the tests run their course. Thanks, @pllim, for thinking this through better!

mhvk commented:
Yay, devdeps is green again! Merging...
